### ISSUE TITLE
Training DenseFusion with another dataset

### ISSUE THREAD

**Issue by BarbeBleue** [j96w/DenseFusion#46]
Hey, 
I developped a database based on a single object, a pink cube, whose data shared the same format than YCB (png images for color, depth and segmentation size 640x480 and .mat file for ground truth values). When I tried to train your network for pose estimation with it by running the .py file with modifications regarding the number of objects and the location of the files to load, most of the time, it failed between the 5th and 15th batch:
```

2019-06-13 15:08:29,614 : Train time 00h 00m 00s Epoch 1 Batch 1 Frame 8 Avg_dis:11.477169811725616
2019-06-13 15:08:29,961 : Train time 00h 00m 01s Epoch 1 Batch 2 Frame 16 Avg_dis:10.867223545908928
2019-06-13 15:08:30,334 : Train time 00h 00m 01s Epoch 1 Batch 3 Frame 24 Avg_dis:14.736500158905983
2019-06-13 15:08:30,701 : Train time 00h 00m 01s Epoch 1 Batch 4 Frame 32 Avg_dis:7.393726512789726
2019-06-13 15:08:31,080 : Train time 00h 00m 02s Epoch 1 Batch 5 Frame 40 Avg_dis:9.500172346830368
2019-06-13 15:08:31,442 : Train time 00h 00m 02s Epoch 1 Batch 6 Frame 48 Avg_dis:11.712907552719116
2019-06-13 15:08:31,813 : Train time 00h 00m 02s Epoch 1 Batch 7 Frame 56 Avg_dis:34.04928183555603
2019-06-13 15:08:32,182 : Train time 00h 00m 03s Epoch 1 Batch 8 Frame 64 Avg_dis:62.857853412628174
2019-06-13 15:08:32,557 : Train time 00h 00m 03s Epoch 1 Batch 9 Frame 72 Avg_dis:12.033220887184143
2019-06-13 15:08:32,917 : Train time 00h 00m 04s Epoch 1 Batch 10 Frame 80 Avg_dis:nan
2019-06-13 15:08:33,348 : Train time 00h 00m 04s Epoch 1 Batch 11 Frame 88 Avg_dis:nan
2019-06-13 15:08:33,703 : Train time 00h 00m 04s Epoch 1 Batch 12 Frame 96 Avg_dis:nan
2019-06-13 15:08:34,071 : Train time 00h 00m 05s Epoch 1 Batch 13 Frame 104 Avg_dis:nan
2019-06-13 15:08:34,489 : Train time 00h 00m 05s Epoch 1 Batch 14 Frame 112 Avg_dis:nan
2019-06-13 15:08:34,865 : Train time 00h 00m 05s Epoch 1 Batch 15 Frame 120 Avg_dis:nan

```
Nan values continue like that until the end.

I have no problem training with YCB the way you did so I think the trouble appeared because I am using an other dataset. 

EDIT:
Looks like the problem could comes from the pred_c values.
At the beginning they are too close from 0. During the loss calculation there is this line:
`loss = torch.mean((dis * pred_c - w * torch.log(pred_c)), dim=0)`
and if at least one pred_c value is considered equal to 0 then torch.log as an -inf value, torch.mean returns inf and during the optimizer step an nan value appear.

However I don't know how to manage this issue, do you have any advice ?

Thank you for your help

**Comment by j96w** [comment]
Hi, nice try! Have you double checked whether the metric is meter? If not, you might need to convert to meter. I think your guessing is correct. Two suggestions might help: (1) Set the pred_c within the range (0.1, 1). If pred_c < 0.1, pred_c = 0.1. (2) Slightly change the hyperparameter w a little bit to see whether it helps avoid the pred_c == 0. I once also had this problem when I change from YCB dataset to Linemod and I solve it by changing the metric to meter and finding a w that works for both dataset. Good luck!

**Comment by BarbeBleue** [comment]
Thank you very much for your answer ! The metric is in meter. The problem is precisely that at the beginning pred_c is far lower than 0.1, for example I have values such as  0.00293281 or even 8.4199555e-21. Increasing w helps, if it is higher than 0.025 the training doesn't stop and the values of pred-c are increasing (really) slowly after the 20th batch.
Could the size of the object affect the result ? Is it too small ?
![001010-color](https://user-images.githubusercontent.com/24253661/59509696-0de5d600-8eb2-11e9-91a0-41023a7798c0.png)

```
{'factor_depth': array([[1000]], dtype=uint16), 'cls_indexes': array([[1]], dtype=uint8), 'instrinsic_matrix': array([[460.,   0., 320.],
       [  0., 460., 240.],
       [  0.,   0.,   1.]]), 'poses': array([[[ 0.23305442],
        [-0.90928744],
        [-0.34479267],
        [-0.11843181]],

       [[-0.85994683],
        [-0.35825158],
        [ 0.36352063],
        [-0.46010978]],

       [[-0.45406726],
        [ 0.21178327],
        [-0.86543097],
        [ 1.66204175]]]), '__globals__': [], '__version__': '1.0', '__header__': b'MATLAB 5.0 MAT-file Platform: posix, Created on: Wed Jun  5 16:01:46 2019'}
```

**Comment by BarbeBleue** [comment]
Just solved the problem ! It was from the way I loaded my masks on the backgrounds now I think it should work as my pred_c values are around 0.9 !

**Comment by j96w** [comment]
The size of the object won't affect the result. It might be better if you add a minimum crop size 20 or 10 into the img crop process. Ok, seems cool but I want to put a notice here that don't let pred_c be too high during the beginning of the training. Since pred_c will gradually increase through the training process, if it starts from 0.9 and grows to 1.0, the argmax step will have some trouble to figure out which per-pixel prediction is the best one. Good luck and nice work~

**Comment by sanjaysswami** [comment]
@BarbeBleue @j96w 
1. I am also trying to train it on my own data but unfortunately I am not able to achieve average distance of less than 12 mm. My training got stuck between 12 mm to 13 mm.
I have tried to increase and decrease parameter w and lr_rate but not getting any good results.

2. My dataset is not real it is synthetic dataset. How much it makes difference any idea?

**Comment by MrLuer** [comment]
> 嘿，
> 我开发了一个基于单个对象的数据库，一个粉红色的立方体，其数据共享的格式与YCB相同（png图像的颜色，深度和分段大小为640x480，.mat文件为地面实况值）。当我尝试通过运行带有关于对象数量和要加载的文件位置的修改的.py文件来训练网络进行姿势估计时，大多数情况下，它在第5批和第15批之间失败：
> 
> ```
> 
> 2019-06-13 15:08:29,614 : Train time 00h 00m 00s Epoch 1 Batch 1 Frame 8 Avg_dis:11.477169811725616
> 2019-06-13 15:08:29,961 : Train time 00h 00m 01s Epoch 1 Batch 2 Frame 16 Avg_dis:10.867223545908928
> 2019-06-13 15:08:30,334 : Train time 00h 00m 01s Epoch 1 Batch 3 Frame 24 Avg_dis:14.736500158905983
> 2019-06-13 15:08:30,701 : Train time 00h 00m 01s Epoch 1 Batch 4 Frame 32 Avg_dis:7.393726512789726
> 2019-06-13 15:08:31,080 : Train time 00h 00m 02s Epoch 1 Batch 5 Frame 40 Avg_dis:9.500172346830368
> 2019-06-13 15:08:31,442 : Train time 00h 00m 02s Epoch 1 Batch 6 Frame 48 Avg_dis:11.712907552719116
> 2019-06-13 15:08:31,813 : Train time 00h 00m 02s Epoch 1 Batch 7 Frame 56 Avg_dis:34.04928183555603
> 2019-06-13 15:08:32,182 : Train time 00h 00m 03s Epoch 1 Batch 8 Frame 64 Avg_dis:62.857853412628174
> 2019-06-13 15:08:32,557 : Train time 00h 00m 03s Epoch 1 Batch 9 Frame 72 Avg_dis:12.033220887184143
> 2019-06-13 15:08:32,917 : Train time 00h 00m 04s Epoch 1 Batch 10 Frame 80 Avg_dis:nan
> 2019-06-13 15:08:33,348 : Train time 00h 00m 04s Epoch 1 Batch 11 Frame 88 Avg_dis:nan
> 2019-06-13 15:08:33,703 : Train time 00h 00m 04s Epoch 1 Batch 12 Frame 96 Avg_dis:nan
> 2019-06-13 15:08:34,071 : Train time 00h 00m 05s Epoch 1 Batch 13 Frame 104 Avg_dis:nan
> 2019-06-13 15:08:34,489 : Train time 00h 00m 05s Epoch 1 Batch 14 Frame 112 Avg_dis:nan
> 2019-06-13 15:08:34,865 : Train time 00h 00m 05s Epoch 1 Batch 15 Frame 120 Avg_dis:nan
> ```
> 
> Nan值继续这样直到结束。
> 
> 我对YCB的培训没有问题，所以我觉得麻烦出现了，因为我正在使用其他数据集。
> 
> 编辑：
> 看起来问题可能来自pred_c值。
> 在开始它们距离0太近。在损失计算期间有这一行：
> `loss = torch.mean((dis * pred_c - w * torch.log(pred_c)), dim=0)`
> 如果至少有一个pred_c值被认为等于0，那么将torch.log作为-inf值，torch.mean返回inf并在优化器期间步出一个纳米值出现。
> 
> 但是我不知道如何处理这个问题，你有什么建议吗？
> 
> 谢谢您的帮助

I want to train on my own dataset，I wonder how you make your dataset？thank you
